[Gatekeep] New cog to catch spam messages from first-time messages from recently joined accounts

Tsunderarislime opened the pull request:
# Gatekeep
This cog attempts to automatically ban accounts that send messages that are considered spam. Of course, with such drastic measures (account ban), this cog targets a very specific subset of users which follow spam bot behaviour.

Users which have joined the server for less than 14 days are placed on a watch list. These users will have their first message sent be evaluated for spam. If it is not spam, they are cleared from the watch list. If it is spam, the user gets banned.

To evaluate the messages for spam, the cog uses a dictionary of words and weights. When a message contains a word in the dictionary, it adds the word's weight to the message's score. If the message's total score exceeds a threshold, the message is flagged as spam.

### Configuration
All of these can be modified using commands. A user *should* need moderator permissions to configure these parameters:
- **Word list**: This is a dictionary of words and weights for message spam evaluation
- **Threshold**: The threshold for which a message must exceed to be considered spam
- **Log channel**: The channel ID to send ban notifications
- **Toggle**: Toggle whether the gatekeeping is active or inactive

### Initial setup
The log channel does need to be assigned, as it defaults to None. The threshold and toggle default values are specifically designed to prevent any accidents. The word list starts empty.

Use `rengatekeep init` to initialize the watch list.

Then just start configuring as needed. Ideally to stop scammers.

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+import logging
+import os
+from random import choice
+import time  # To auto remove birthday role on the next day.
```
Nice, can you update the docstring/comments?

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+    @checks.mod_or_permissions(administrator=True)
+    async def deactivate(self, ctx: Context):
+        """ Activate the gatekeeping cog
+        """
```
1. `Deactivate`
2. Can you put this docstring in a single line?

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+    @checks.mod_or_permissions(administrator=True)
+    async def activate(self, ctx: Context):
+        """ Activate the gatekeeping cog
+        """
```
nit: Here and anywhere else a docstring is a single line: please keep the triple quotes on the same line.

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+    @checks.mod_or_permissions(administrator=True)
+    async def addWord(self, ctx: Context, word: str, weight: int):
+        """Add a word to the gatekeeping list. If the word already exists on the
+        list, then update the word weight to the new weight.
```
Can you put `If the word already...` and beyond on a new line? The first line of the docstring is used in the help menu when you pass it only `[p]gatekeep`.

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+                )
+
+                # Do the deed
+                await message.author.ban(delete_message_days=7)
```
Can you please pass in a `reason` for the audit log?

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+            if score >= th:
+                # Proceed to ban and announce to mod log channel
+                ch = await self.config.guild(message.guild).get_attr(KEY_LOG_CHANNEL)()
+                await self.bot.get_channel(ch).send(f'Banned <@{author.id}> `{author.id}` for posting spam. The message score was {score}, which exceeded the threshold of {th}. Their message was:\n `{message.content}`')
```
1. Use `{author.mention}` to construct the mention.
2. Can you also guard against the max character limit of 2k by truncating it with `...`?

Tsunderarislime replied:
![image](https://github.com/user-attachments/assets/f800abcf-a371-47b5-bc8d-b5390a7a14d9)
300 messages larger than 300 characters now get truncated, change if needed. Updated mention to use author.mention

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,499 @@
+
+    @commands.group(name="gatekeep")
+    @commands.guild_only()
+    async def _gatekeep(self, ctx: Context):
```
Can you also decorate with `@checks.mod_or_permissions(administrator=True)`? want to avoid exposing this command tree to non-mods/admin.

Injabie3 reviewed (changes requested):
Thanks for doing this, I did a first pass.

Could you also please rebase your changes onto the tip of V3/testing so that we can reduce the diff size to your actual changes?

Tsunderarislime commented:
New changes since the first commit and rebasing hell:

- Number of days that a user needs to be in the server can now be configured `rengatekeep days <days>`
- Removed BASE_GUILD_MEMBER from **constants.py**, this was a carryover from using the birthday cog as a template
- Reordered ban notification. Spammers no longer get mentioned _before_ they get banned. This likely doesn't actually change anything if the spammers can't see the mod log channel, but better safe than sorry
- Removed a few unused imports in **gatekeep.py**

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,526 @@
+                str(threshold),
+            )
+            await ctx.send(
+                ":white_check_mark: **Gatekeep - Threshold**: The threshold has been updated to **{}**".format(threshold)
```
Here and everywhere else `.format` is used: can you also use f-strings to be consistent across the file

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,526 @@
+        active = await self.config.guild(ctx.guild).get_attr(KEY_ACTIVE)()
+        threshold = await self.config.guild(ctx.guild).get_attr(KEY_THRESHOLD)()
+        nDays = await self.config.guild(ctx.guild).get_attr(KEY_NEW_USER_DAYS)()
+        await ctx.send(f":information_source: Current Status :information_source:\n- Log Channel: <#{log}>\n- Gatekeeping: {active}\n- Threshold: {threshold}\n- Days to watch: {nDays}")
```
nit: Can you put this into a variable and split this across multiple lines?

Tsunderarislime replied:
Not sure what you mean by put into a variable, but the string is now multiple lines instead of one really long line

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,526 @@
+        try:
+            response = await self.bot.wait_for("message", timeout=30.0, check=check)
+        except asyncio.TimeoutError:
+            await ctx.send(f"No response after 30 seconds, this operation will not be executed.")
```
no need for `f` since there's no variables.

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,526 @@
+
+        Parameters:
+        -----------
+        threshold: Integer
```
Here and elsewhere with `Integer` and `String`: can you use the type name in the docstring? e.g. `int` and `str`?

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,526 @@
+    @_gatekeep.command(name="add")
+    @commands.guild_only()
+    @checks.mod_or_permissions(administrator=True)
+    async def addWord(self, ctx: Context, word: str, weight: int):
+        """Add a word to the gatekeeping list.
+        If the word already exists on the list,
+        then update the word weight to the new weight.
+
+        Parameters:
+        -----------
+        word: String
+            The word to be added to the gatekeep list. These will automatically be
+            converted to lowercase.
+
+        weight: Integer
+            The weight that the word has in the gatekeep list. Higher weights mean
+            the word is more likely to flag the entire message as spam.
+        """
+
+        # Ensure both the word is non-empty and the weight is greater than 0
+        if word and weight > 0:
```
Here and in removeWord: Can you sanitize the `word` input with a helper before we use it, perhaps [redbot.core.utils.chat_formatting.escape](https://docs.discord.red/en/stable/framework_utils.html#redbot.core.utils.chat_formatting.escape)?

Tsunderarislime replied:
Not too sure how much sanitation needed, but right now it removes every character in Python's string.punctuation constant. This constant does include the entire set of characters used in markdown.

See this example below:
![image](https://github.com/user-attachments/assets/5d4c54ed-24ad-4e73-91bb-a640977f75b8)
The words in the filter at the time was `everyone`, `free`, and `playstation`. These had weights of `6`, `3`, and `3` respectively.

Injabie3 commented on cogs/gatekeep/gatekeep.py:
```diff
@@ -0,0 +1,526 @@
+        if not valid_user:
+            return
+
+        if await self.bot.is_automod_immune(message):
+            return
```
Do you know if this also makes admins/mod/bot owners immune?

Tsunderarislime replied:
The `is_automod_immune(...)` method is part of RedBot's core, as seen below in this snippet.
https://github.com/SFUAnime/Ren/blob/c83643461671d9a8011f36769d2ebd0aabd623be/redbot/core/bot.py#L1894-L1936

Roles IDs need to be added to a server's config in order to have the method evaluate to true. See below example:

![image](https://github.com/user-attachments/assets/56adb09d-8943-4738-b771-1eaac7412739)

Funnily enough, roles with administrator permissions **do not** count as automod immune by default. The bot did successfully ban my test user even when they had a role that grants administrator permissions.

I added a quick fix to make those with administrator permission immune. I also added those checks in the init method and the daily list refresh for redundancy. Better to be safe than sorry.

Injabie3 commented on cogs/gatekeep/info.json:
```diff
@@ -0,0 +1,10 @@
+    "name": "Gatekeep",
+    "short": "Automatically ban scammers that send messages with user-defined keywords",
+    "description": "Automatically detects and bans scammers that send messages containing user-defined keywords.",
+    "end_user_data_statement": "This cog will store users' Discord ID of those who joined the server less than 14 days ago.",
```
Can you remove the 14 day part since this is user configurable? I think leaving it generic with "This cog will store users' Discord IDs" is sufficient.

Injabie3 reviewed (changes requested):
Generally the logic LGTM, some additional comments below.

Tsunderarislime commented:
![image](https://github.com/user-attachments/assets/4db49b5b-ab1d-442b-9676-4c88967eaf30)
![image](https://cdn.discordapp.com/emojis/1183273806832345199.webp?size=32) ![image](https://cdn.discordapp.com/emojis/1183273806832345199.webp?size=32) ![image](https://cdn.discordapp.com/emojis/1183273806832345199.webp?size=32)

Injabie3 reviewed (approved):
Thanks, the general functionality LGTM!

Could you make another PR that adds a `user` command ~subtree~ group with `add`/`remove`/`list`? This will also mean we'll want a `word` command ~subtree~ group with the same to avoid conflict.

Also, may I have your permission to cherry-pick these changes to the [Injabie3/lui-cogs-v3][1] repo? We'll need it there to deploy to Ren.

[1]: https://github.com/Injabie3/lui-cogs-v3

Injabie3 commented:
@quachtridat if you have additional feedback, we can address in subsequent PR.
